### PR TITLE
[QUIC] Certificate name validation

### LINKED ISSUE #55193: [HTTP/3] Certificate name validation different than HTTP/1.1?
I have a Kestrel server configured to serve HTTP/1.1 and HTTP/3 on the same port with the same certificate. The certificate is the ASP.NET Core development certificate.

`HttpClient` throws this error when calling with HTTP/1.1:
```
    System.Net.Http.HttpRequestException : The SSL connection could not be established, see inner exception.
    ---- System.Security.Authentication.AuthenticationException : The remote certificate is invalid according to the validation procedure: RemoteCertificateNameMismatch
```
I need to setup a handler with `ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator` to successfully call it.

Meanwhile, `HttpClient` calling HTTP/3 doesn't need the validation callback. It happily accepts the dev certificate as trusted.

Should HTTP/1.1 and HTTP/2 have different certificate validation behavior than HTTP/3?

### PR DESCRIPTION
**Pull request by ManickaP**
Extracts `CertificateValidation` for Windows into shared sources and make use of it and the Unix version in S.N.Quic to properly verify the cert the same way `SslStream` does.

Fixes #55193

For now just manually verified in asp.net core test. I will add specific tests tomorrow.

cc: @wfurt

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 1]
```diff
@@ -39,6 +39,14 @@ internal sealed class MsQuicConnection : QuicConnectionProvider
         private bool _remoteCertificateRequired;
         private X509RevocationMode _revocationMode = X509RevocationMode.Offline;
         private RemoteCertificateValidationCallback? _remoteCertificateValidationCallback;
+        private string? _targetHost;
+
+        private string TargetHost => _targetHost ?? (_targetHost = _remoteEndPoint switch
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:44]: @wfurt I needed the same thing as you in #55468. We'll clash on this and one of us will have to fix their part.

**Comment by wfurt** [comment]
with this, validation _should_ be easier in ldap @AnthonyMastrean (#55017)

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx [hunk 2]
```diff
@@ -150,5 +150,8 @@
   <data name="net_quic_writing_notallowed" xml:space="preserve">
     <value>Writing is not allowed on stream.</value>
   </data>
+  <data name="net_ssl_app_protocols_invalid" xml:space="preserve">
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/Resources/Strings.resx:153]: do we need this? looks like extra.

File: src/libraries/System.Net.Quic/src/System.Net.Quic.csproj [hunk 3]
```diff
@@ -83,6 +121,14 @@
     <Reference Include="System.Threading.Channels" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="$(LibrariesProjectRoot)System.Security.Cryptography.OpenSsl\src\System.Security.Cryptography.OpenSsl.csproj" />
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System.Net.Quic.csproj:125]: Do we still need this with all the added interop?

File: src/libraries/System.Net.Quic/src/System.Net.Quic.csproj [hunk 4]
```diff
@@ -83,6 +121,14 @@
...
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetsUnix)' == 'true'">
+    <Reference Include="System.Diagnostics.StackTrace" />
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System.Net.Quic.csproj:129]: Extra? If anything I would add it to the Debug only addition for Console. perhaps as separate PR?

**Review by wfurt — COMMENTED** [review]
generally looks god to me. We should add some testes.
- set TargetHost to something else than "loopback" with existing cert and the name error should show up.
- In ideal case we should check that connection to 127.0.0.1 and ::1 works without getting the name error since they both should be in altName (at least the one we generate) 
I think current tests ignore all the passed info and simple call new chain rebuild without name check.

File: src/libraries/System.Net.Quic/src/Resources/Strings.resx (hunk 2 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/Resources/Strings.resx:153]: Yes we need this, it's used by one of the included interop source files:
https://github.com/dotnet/runtime/blob/d148c34bdfbd49e3d820eaa4dcbc47832d52d0d2/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.SslCtx.cs#L43-L43

https://github.com/dotnet/runtime/pull/56175/files#diff-1d4ce147edca0ce81f312b96fce76ac52eb02eb6d04e660817e3b47230edc794R64

File: src/libraries/System.Net.Quic/src/System.Net.Quic.csproj (hunk 3 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System.Net.Quic.csproj:125]: Yes, we do, the interop uses public classes from this project. It doesn't build without it and the same it's referenced in S.N.Http.

File: src/libraries/System.Net.Quic/src/System.Net.Quic.csproj (hunk 4 above)
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System.Net.Quic.csproj:129]: Not extra, used by one the added interop sources:
https://github.com/dotnet/runtime/blob/b937677e8f8601848d29bc072a93cc0c6e21576d/src/libraries/Common/src/Microsoft/Win32/SafeHandles/SafeX509Handles.Unix.cs#L16-L16

But you're right it's only for Debug, I'll fix that.

Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System.Net.Quic.csproj]: Fixed.

**Comment by ManickaP** [comment]
@wfurt tests added as you suggested. If you're happy with it, feel free to merge on my behalf.

**Review by wfurt — APPROVED** [review]
LGTM

**Pull request merged**